File: src/test/java/com/rafaelfiume/tictactoe/console/acceptance/AbstractConsoleTicTacToeTest.java
package com.rafaelfiume.tictactoe.console.acceptance;

import com.googlecode.yatspec.junit.SpecResultListener;
import com.googlecode.yatspec.junit.WithCustomResultListeners;
import com.googlecode.yatspec.plugin.sequencediagram.SequenceDiagramGenerator;
import com.googlecode.yatspec.plugin.sequencediagram.SvgWrapper;
import com.googlecode.yatspec.rendering.html.DontHighlightRenderer;
import com.googlecode.yatspec.rendering.html.HtmlResultRenderer;
import com.googlecode.yatspec.rendering.html.index.HtmlIndexRenderer;
import com.googlecode.yatspec.state.givenwhenthen.ActionUnderTest;
import com.googlecode.yatspec.state.givenwhenthen.GivensBuilder;
import com.googlecode.yatspec.state.givenwhenthen.StateExtractor;
import com.googlecode.yatspec.state.givenwhenthen.TestState;
import com.rafaelfiume.tictactoe.Board;
import com.rafaelfiume.tictactoe.console.ConsoleGameRunner;
import com.rafaelfiume.tictactoe.support.ConsoleInputReaderStub;
import com.rafaelfiume.tictactoe.support.RecordConsoleOutputRenderer;

import static com.googlecode.totallylazy.Sequences.sequence;

public abstract class AbstractConsoleTicTacToeTest extends TestState implements WithCustomResultListeners {

    // TODO RF 26/11/15 Proposal to extract Turn from Board...
    //private final Board board = new Board();
    //private final Turn turn = mock(turn);
    //private final TicTacToe ticTacToe = new TicTacToe(board, turn);

    private final ConsoleInputReaderStub consoleInputReader = new ConsoleInputReaderStub();
    private final RecordConsoleOutputRenderer consoleRenderer = new RecordConsoleOutputRenderer();
    private final ConsoleGameRunner gameRunner = new ConsoleGameRunner(new Board(), consoleInputReader, consoleRenderer);

    @Override
    public Iterable<SpecResultListener> getResultListeners() throws Exception {
        return sequence(
                new HtmlResultRenderer().
                        withCustomHeaderContent(SequenceDiagramGenerator.getHeaderContentForModalWindows()).
                        withCustomRenderer(SvgWrapper.class, new DontHighlightRenderer()),
                new HtmlIndexRenderer()).
                safeCast(SpecResultListener.class);
    }

    protected ActionUnderTest player_X_marksTopLeftPosition() { return userHitsNumber(1); }
    protected ActionUnderTest player_O_marksTopLeftPosition() { return userHitsNumber(1); }
    protected ActionUnderTest player_X_marksTopCenterPosition() { return userHitsNumber(2); }
    protected ActionUnderTest player_O_marksTopCenterPosition() { return userHitsNumber(2); }
    protected ActionUnderTest player_X_marksTopRightPosition() { return userHitsNumber(3); }
    protected ActionUnderTest player_X_marksMidLeftPosition() { return userHitsNumber(4);}
    protected ActionUnderTest player_O_marksMidLeftPosition() { return userHitsNumber(4); }
    protected ActionUnderTest player_X_marksCenterPosition() { return userHitsNumber(5); }
    protected ActionUnderTest player_O_marksCenterPosition() { return userHitsNumber(5); }
    protected ActionUnderTest player_X_marksMidRightPosition() { return userHitsNumber(6); }
    protected ActionUnderTest player_O_marksMidRightPosition() { return userHitsNumber(6); }
    protected ActionUnderTest player_X_marksDownLeftPosition() { return userHitsNumber(7); }
    protected ActionUnderTest player_O_marksDownLeftPosition() { return userHitsNumber(7); }
    protected ActionUnderTest player_X_marksDownCenterPosition() { return userHitsNumber(8); }
    protected ActionUnderTest player_X_marksDownRightPosition() { return userHitsNumber(9); }
    protected ActionUnderTest player_O_marksDownRightPosition() { return userHitsNumber(9); }

    protected ActionUnderTest player_O_entersInvalidCellInTheBoard() {
        return userHitsNumber(99);
    }

    protected void andAppIsUpAndRunning() {
        startAppWithAHack();
    }

    protected GivensBuilder appIsUpAndRunning() {
        return givens -> {
            startAppWithAHack();
            return givens;
        };
    }

    protected TestState given(ActionUnderTest actionUnderTest) throws Exception {
        // Bypassing Yatspec framework since its original given implementation is not helping much in these tests (tests are reading way better now)
        return when(actionUnderTest);
    }

    protected TestState and(ActionUnderTest actionUnderTest) throws Exception {
        return when(actionUnderTest);
    }

    protected StateExtractor<Object> theGameDescription() {
        return inputAndOutputs -> consoleRenderer.gameDescription();
    }

    protected StateExtractor<String> theGame() {
        return inputAndOutputs -> consoleRenderer.boarGame();
    }

    protected StateExtractor<String> theGameStatus() {
        return inputAndOutputs -> consoleRenderer.gameStatus();
    }

    private ActionUnderTest userHitsNumber(int number) {
        return (givens, capturedInputAndOutputs) -> {
            consoleInputReader.willReturn(number);
            return capturedInputAndOutputs;
        };
    }

    private void startAppWithAHack() {
        try {
            gameRunner.start();
        } catch (NullPointerException e) {
            // Test and app work all right, but...
            // This is a massive hack to deal with premature termination: game is not over an ConsoleInputReader#readUserInput will blow up
            // TODO RF 25/11/2015 Improve the design to fix it
        }
    }

}

File: src/test/java/com/rafaelfiume/tictactoe/acceptance/AbstractConsoleTicTacToeTest.kt
package com.rafaelfiume.tictactoe.acceptance

import com.googlecode.totallylazy.Sequences.sequence
import com.googlecode.yatspec.junit.SpecResultListener
import com.googlecode.yatspec.junit.WithCustomResultListeners
import com.googlecode.yatspec.plugin.sequencediagram.SequenceDiagramGenerator
import com.googlecode.yatspec.plugin.sequencediagram.SvgWrapper
import com.googlecode.yatspec.rendering.html.DontHighlightRenderer
import com.googlecode.yatspec.rendering.html.HtmlResultRenderer
import com.googlecode.yatspec.rendering.html.index.HtmlIndexRenderer
import com.googlecode.yatspec.state.givenwhenthen.ActionUnderTest
import com.googlecode.yatspec.state.givenwhenthen.GivensBuilder
import com.googlecode.yatspec.state.givenwhenthen.StateExtractor
import com.googlecode.yatspec.state.givenwhenthen.TestState
import com.rafaelfiume.tictactoe.TttGame.Factory.newGame
import com.rafaelfiume.tictactoe.console.ConsoleGameRunner
import com.rafaelfiume.tictactoe.support.ConsoleInputReaderStub
import com.rafaelfiume.tictactoe.support.RecordConsoleOutputRenderer

abstract class AbstractConsoleTicTacToeTest : TestState(), WithCustomResultListeners {

    private val consoleInputReader = ConsoleInputReaderStub()
    private val consoleRenderer = RecordConsoleOutputRenderer()
    private val gameRunner = ConsoleGameRunner(newGame(), consoleInputReader, consoleRenderer)

    @Throws(Exception::class)
    override fun getResultListeners(): Iterable<SpecResultListener> {
        return sequence(
                HtmlResultRenderer().withCustomHeaderContent(SequenceDiagramGenerator.getHeaderContentForModalWindows()).withCustomRenderer(SvgWrapper::class.java, DontHighlightRenderer()),
                HtmlIndexRenderer()).safeCast(SpecResultListener::class.java)
    }

    protected fun player_X_types_1(): ActionUnderTest = userHitsNumber(1)
    protected fun player_X_types_2(): ActionUnderTest = userHitsNumber(2)
    protected fun player_X_types_3(): ActionUnderTest = userHitsNumber(3)
    protected fun player_X_types_4(): ActionUnderTest = userHitsNumber(4)
    protected fun player_X_types_5(): ActionUnderTest = userHitsNumber(5)
    protected fun player_X_types_6(): ActionUnderTest = userHitsNumber(6)
    protected fun player_X_types_7(): ActionUnderTest = userHitsNumber(7)
    protected fun player_X_types_8(): ActionUnderTest = userHitsNumber(8)
    protected fun player_X_types_9(): ActionUnderTest = userHitsNumber(9)

    protected fun player_O_types_1(): ActionUnderTest = userHitsNumber(1)
    protected fun player_O_types_2(): ActionUnderTest = userHitsNumber(2)
    protected fun player_O_types_4(): ActionUnderTest = userHitsNumber(4)
    protected fun player_O_types_5(): ActionUnderTest = userHitsNumber(5)
    protected fun player_O_types_6(): ActionUnderTest = userHitsNumber(6)
    protected fun player_O_types_7(): ActionUnderTest = userHitsNumber(7)
    protected fun player_O_types_9(): ActionUnderTest = userHitsNumber(9)

    protected fun player_O_typesInvalidStuff(): ActionUnderTest = userHitsNumber(99)

    protected fun andAppIsUpAndRunning() {
        startAppWithAHack()
    }

    protected fun appIsUpAndRunning() = GivensBuilder { givens ->
        startAppWithAHack()
        givens
    }

    @Throws(Exception::class)
    // Bypassing Yatspec framework since its original given implementation is not helping much in these tests (tests are reading way better now)
    protected fun given(actionUnderTest: ActionUnderTest): TestState = `when`(actionUnderTest)

    @Throws(Exception::class)
    protected fun and(actionUnderTest: ActionUnderTest): TestState = `when`(actionUnderTest)

    protected fun theGameDescription() = StateExtractor { inputAndOutputs -> consoleRenderer.gameDescription() }
    protected fun theGame()            = StateExtractor { inputAndOutputs -> consoleRenderer.boarGame()        }
    protected fun theGameStatus()      = StateExtractor { inputAndOutputs -> consoleRenderer.gameStatus()      }

    private fun userHitsNumber(number: Int) = ActionUnderTest { givens, capturedInputAndOutputs ->
        consoleInputReader.willReturn(number)
        capturedInputAndOutputs
    }

    private fun startAppWithAHack() {
        try {
            gameRunner.start()
        } catch (e: NullPointerException) {
            // Test and app work all right, but...
            // TODO RF 25/11/2015 This is a massive hack to deal with premature termination: game is not over an ConsoleInputReader#readUserInput will blow up
        }
    }

}
